### PR TITLE
nightly GHA job

### PR DESCRIPTION
**Pull request by Ginxo**
### JIRA
https://issues.redhat.com/browse/PLANNER-2604

Can you please guys fill the gaps? from `Set failure message`, `Set success message` and `Send a stream message` steps?

Once we have it clear I will create `ZULIP_API_KEY` secret before merging.
I set two different event types:

- `schedule` to be run based on cron expression `0 0 * * *` (every night)(for default branch, stable in this case)
- `workflow_dispatch` to manually run it (choosing the branch)

You can check execution tests here https://github.com/Ginxo/optaplanner-quickstarts/actions/workflows/nightly.yml

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>full downstream build</b> please add the label `run_fdb`
* for a <b>specific pull request build</b> please add comment: <b>Jenkins (re)run [optaplanner-quickstarts] tests</b>
* for a <b>Quarkus LTS check</b> please add comment: <b>Jenkins run LTS</b>
* for a <b>specific Quarkus LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner-quickstarts] LTS</b>
* for a <b>Native check</b> please add comment: <b>Jenkins run native</b>
* for a <b>specific Native LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner-quickstarts] native</b>
</details>

### CODE REVIEW

File: .github/workflows/nightly.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+jobs:
+  nightly:
+    concurrency:
+      group: nightly
+      cancel-in-progress: true
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        java-version: [11]
```
Comment (Ginxo) [inline .github/workflows/nightly.yml:17]: is java11 the proper one? 17 maybe?

File: .github/workflows/nightly.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+on:
+  schedule:
+    - cron: "0 0 * * *"
+  workflow_dispatch:
+
+jobs:
+  nightly:
+    concurrency:
+      group: nightly
```
Comment (Ginxo) [inline .github/workflows/nightly.yml:12]: I'm forcing the nightly to not have concurrency executions

File: .github/workflows/nightly.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
+
+name: Nightly
+
+on:
+  schedule:
+    - cron: "0 0 * * *"
```
Comment (Ginxo) [inline .github/workflows/nightly.yml:6]: I guess you don't care exactly the time to run the job so I decided 00:00, do you agree?

File: .github/workflows/nightly.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        if: ${{ failure() }}
+        run: echo "zulip_message='The content of the message. Maximum message size of 10000 bytes'" >> $GITHUB_ENV  
+      - name: Set success message
+        if: ${{ success() }}
+        run: echo "zulip_message='The content of the message. Maximum message size of 10000 bytes'" >> $GITHUB_ENV  
+      - name: Send a stream message
+        if: ${{ always() }}
+        uses: zulip/github-actions-zulip/send-message@v1
+        with:
+          api-key: ${{ secrets.ZULIP_API_KEY }}
```
Comment (Ginxo) [inline .github/workflows/nightly.yml:50]: don't forge to create `ZULIP_API_KEY` secret before merging

File: .github/workflows/nightly.yml (hunk 1 above)
Comment (rsynek) [reply .github/workflows/nightly.yml:17]: Any of these is fine. The goal is to make sure the `stable` branch was not corrupted during the release. Either directly or indirectly, e.g. by not having the artifacts in the Maven Central yet.

File: .github/workflows/nightly.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        with:
+          java-version: ${{ matrix.java-version }}
+          maven-version: ${{ matrix.maven-version }}
+          cache-key-prefix: ${{ runner.os }}-${{ matrix.java-version }}-maven${{ matrix.maven-version }}
+      - uses: actions/checkout@v2      
+      - name: Build with Maven
+        run: mvn -B clean install --file pom.xml dependency:tree -Dmaven.wagon.httpconnectionManager.ttlSeconds=25 -Dmaven.wagon.http.retryHandler.count=3
+      - name: Set failure message
+        if: ${{ failure() }}
+        run: echo "zulip_message='The content of the message. Maximum message size of 10000 bytes'" >> $GITHUB_ENV
```
Comment (rsynek) [inline .github/workflows/nightly.yml]: Is it possible to also include a link to the failed job run?
```suggestion
        run: echo "zulip_message='The optaplanner-quickstarts stable branch has failed.'" >> $GITHUB_ENV  
```

File: .github/workflows/nightly.yml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+          cache-key-prefix: ${{ runner.os }}-${{ matrix.java-version }}-maven${{ matrix.maven-version }}
+      - uses: actions/checkout@v2      
+      - name: Build with Maven
+        run: mvn -B clean install --file pom.xml dependency:tree -Dmaven.wagon.httpconnectionManager.ttlSeconds=25 -Dmaven.wagon.http.retryHandler.count=3
+      - name: Set failure message
+        if: ${{ failure() }}
+        run: echo "zulip_message='The content of the message. Maximum message size of 10000 bytes'" >> $GITHUB_ENV  
+      - name: Set success message
+        if: ${{ success() }}
+        run: echo "zulip_message='The content of the message. Maximum message size of 10000 bytes'" >> $GITHUB_ENV
```
Comment (rsynek) [inline .github/workflows/nightly.yml]:
```suggestion
        run: echo "zulip_message='The optaplanner-quickstarts stable branch was successfully tested.'" >> $GITHUB_ENV  
```

File: .github/workflows/nightly.yml [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      - name: Set success message
+        if: ${{ success() }}
+        run: echo "zulip_message='The content of the message. Maximum message size of 10000 bytes'" >> $GITHUB_ENV  
+      - name: Send a stream message
+        if: ${{ always() }}
+        uses: zulip/github-actions-zulip/send-message@v1
+        with:
+          api-key: ${{ secrets.ZULIP_API_KEY }}
+          email: 'Email address of the user who owns the API key mentioned above'
+          organization-url: 'Zulip organization canonical URL'
```
Comment (rsynek) [inline .github/workflows/nightly.yml]:
```suggestion
          organization-url: 'https://kie.zulipchat.com'
```

File: .github/workflows/nightly.yml [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+          to: 'the name or integer ID of the stream'
```
Comment (rsynek) [inline .github/workflows/nightly.yml]:
```suggestion
          to: 'optaplanner-dev'
```

File: .github/workflows/nightly.yml [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+          type: 'private for a private message and stream for a stream message'
```
Comment (rsynek) [inline .github/workflows/nightly.yml]:
```suggestion
          type: 'stream'
```

File: .github/workflows/nightly.yml [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+          topic: 'Only required for stream messages (type="stream"), ignored otherwise'
```
Comment (rsynek) [inline .github/workflows/nightly.yml]:
```suggestion
          topic: 'status of optaplanner-quickstarts - stable'
```

**Review by rsynek — APPROVED** [review]

File: .github/workflows/nightly.yml (hunk 5 above)
Comment (Ginxo) [reply .github/workflows/nightly.yml]:
```suggestion
        run: echo "zulip_message='The ${{github.repository}} ${GITHUB_REF#refs/heads/} branch has failed.'" >> $GITHUB_ENV  
```

File: .github/workflows/nightly.yml (hunk 6 above)
Comment (Ginxo) [reply .github/workflows/nightly.yml]:
```suggestion
        run: echo "zulip_message='The ${{github.repository}} ${GITHUB_REF#refs/heads/} branch was successfully tested.'" >> $GITHUB_ENV  
```

File: .github/workflows/nightly.yml (hunk 10 above)
Comment (Ginxo) [reply .github/workflows/nightly.yml]:
```suggestion
          topic: 'status of ${{github.repository}} - ${GITHUB_REF#refs/heads/} '
```

File: .github/workflows/nightly.yml [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        run: echo "zulip_message='The ${{github.repository}} ${GITHUB_REF#refs/heads/} branch has failed.'" >> $GITHUB_ENV  
+      - name: Set success message
+        if: ${{ success() }}
+        run: echo "zulip_message='The ${{github.repository}} ${GITHUB_REF#refs/heads/} branch was successfully tested.'" >> $GITHUB_ENV  
+      - name: Send a stream message
+        if: ${{ always() }}
+        uses: zulip/github-actions-zulip/send-message@v1
+        with:
+          api-key: ${{ secrets.ZULIP_API_KEY }}
+          email: 'Email address of the user who owns the API key mentioned above'
```
Comment (Ginxo) [inline .github/workflows/nightly.yml]: this still pending @rsynek @Ginxo

File: .github/workflows/nightly.yml [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+          organization-url: 'https://kie.zulipchat.com'
+          to: 'optaplanner-dev'
+          type: 'stream'
+          topic: 'status of ${{github.repository}} - ${GITHUB_REF#refs/heads/} '
```
Comment (Ginxo) [inline .github/workflows/nightly.yml]:
```suggestion
          topic: 'status of ${{github.repository}} - ${GITHUB_REF#refs/heads/}'
```

File: .github/workflows/nightly.yml [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        run: echo "zulip_message='The ${{github.repository}} ${GITHUB_REF#refs/heads/} branch has failed.'" >> $GITHUB_ENV  
+      - name: Set success message
+        if: ${{ success() }}
+        run: echo "zulip_message='The ${{github.repository}} ${GITHUB_REF#refs/heads/} branch was successfully tested.'" >> $GITHUB_ENV  
+      - name: Send a stream message
+        if: ${{ always() }}
+        uses: zulip/github-actions-zulip/send-message@v1
+        with:
+          api-key: ${{ secrets.ZULIP_API_KEY }}
+          email: ${{ secrets.ZULIP_EMAIL }}
```
Comment (Ginxo) [inline .github/workflows/nightly.yml:51]: `ZULIP_EMAIL` secret, so don't foget to create it before merging

File: .github/workflows/nightly.yml [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+      - name: Build with Maven
+        run: mvn -B clean install --file pom.xml dependency:tree -Dmaven.wagon.httpconnectionManager.ttlSeconds=25 -Dmaven.wagon.http.retryHandler.count=3
+      - name: Set build information
+        if: ${{ always() }}
+        run: |
+          echo "project_branch=${GITHUB_REF#refs/heads/}" >> $GITHUB_ENV  
+          echo "project_name=${{github.repository}}" >> $GITHUB_ENV  
+      - name: Set failure message
+        if: ${{ failure() }}
+        run: echo "zulip_message='The ${{env.project_name}} ${{env.project_branch}} branch has failed.'" >> $GITHUB_ENV
```
Comment (Ginxo) [inline .github/workflows/nightly.yml]: @rsynek seems redundant... wdyt?
image

Comment (Ginxo) [reply .github/workflows/nightly.yml]: see my proposal from https://github.com/kiegroup/optaplanner-quickstarts/pull/263/commits/45dfefec897ef4b431a2ce270d4c181b8d342469
image

Comment (rsynek) [reply .github/workflows/nightly.yml]: Good point.

**Review by rsynek — APPROVED** [review]

**Comment by Ginxo** [comment]
@rsynek @ge0ffrey can you please create `ZULIP_EMAIL` and `ZULIP_API_KEY` secrets?

- go https://github.com/kiegroup/optaplanner-quickstarts/settings/secrets/actions
- click on `New Repository Secret`
- save

to get `ZULIP_API_KEY`

- go to https://kie.zulipchat.com/#settings/account-and-privacy
- Show/change your API key

**Comment by rsynek** [comment]
> ZULIP_EMAIL

Done. I assume that the `ZULIP_EMAIL`  is the email address for `optaplanner-dev` stream.

**Comment by Ginxo** [comment]
> > ZULIP_EMAIL
> 
> Done. I assume that the `ZULIP_EMAIL` is the email address for `optaplanner-dev` stream.

whatever the email you use (I guess) should match with the one to get the API key

**Comment by Ginxo** [comment]
moved to https://github.com/kiegroup/optaplanner-quickstarts/pull/265

**Pull request closed without merging**